### ISSUE TITLE
EPIC: Make optional dependencies available via conda-forge

### ISSUE THREAD

**Issue by corneliusroemer** [conda-forge/polars-feedstock#244]
### Comment:

I noticed that some optional polars dependencies aren't available through conda-forge. It would be nice if all optional dependencies could be installed via conda - so that one doesn't need pip.

Missing (to the best of my knowledge):

- [ ] xlsx2csv
- [ ] deltalake

Also, it would be good to add the version constraints from https://github.com/pola-rs/polars/blob/master/py-polars/pyproject.toml#L42-L59 as run-constraints to the polars conda-forge recipe

*transferred from https://github.com/pola-rs/polars/issues/7585*

**Comment by 0xbe7a** [comment]
`xlsx2csv` and `deltalake` are both available on conda-forge and are nearly up-to-date (https://github.com/conda-forge/xlsx2csv-feedstock/pull/1 @borchero 👀).

I am not sure how we can map pip's optional dependencies to conda until https://github.com/conda/ceps/pull/55 is accepted, without introducing `run_constraints` for the package in general.

**Comment by borchero** [comment]
I recently noticed that `fastexcel` is missing from `conda-forge` though, I'll take care in the next couple days.

**Comment by borchero** [comment]
I merged the `xlsx2csv` version update, thanks for pointing me to it @0xbe7a

**Comment by corneliusroemer** [comment]
Added `fastexcel` to the todo-list :)

**Comment by corneliusroemer** [comment]
If I understand run constraints correctly, they can be optional - that's the whole point? That they are not requirements but constraints in case the optional requirement is fulfilled?

@0xbe7a maybe I misunderstand, but to me that's what the docs say pretty explicitly.

See: https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#run-constrained

E.g.

> ### Run_constrained
> Packages that are optional at runtime but must obey the supplied additional constraint if they are installed.
> 
> Package names should follow the [package match specifications](https://conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#package-match-specifications).
> 
> ```
> requirements:
>   run_constrained:
>     - optional-subpackage =={{ version }}
>  ```

and

> If build and link dependencies need to impose constraints on the run environment but not necessarily pull in additional packages, then this can be done by altering the [Run_constrained](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#run-constrained) entries. In addition to weak/strong run_exports which add to the run requirements, weak_constrains and strong_constrains add to the run_constrained requirements. With these, e.g., minimum versions of compatible but not required packages (like optional plugins for the linked dependency, or certain system attributes) can be expressed:

**Comment by 0xbe7a** [comment]
However, the `run_constraint` is also enforced even if the user is not using the Polars features that depend on these packages. For example, we might have a large environment containing `openpyxl 2.*` because some other logic makes use of it and is not using the Polars "read excel" functionality. The `run_constraint` would still enforce `openpyxl 3.*`. Conda would fail to solve this case, while with pip, we could simply not enable the optional feature.

**Comment by corneliusroemer** [comment]
True, but no package truly always use all the functionality that give rise to a particular constraint. I guess it depends how commonly used the optional dependency is and how tightly you are constrained/constraining. It's always a tradeoff.

**Comment by 0xbe7a** [comment]
> The best way to ensure solve issues would be to ensure optional dependencies are as little constrained as possible.

I see your point, and I also think that this should be a tradeoff based on the popularity of the dependency. I am hesitant here, as there is nothing the end-user can do to ignore these potentially "bogus" constraints and create the environment regardless of any constraints. Feel free to open a PR for any potential constraints. I think it's much better to go through this on a case-by-case basis rather than in general.

**Comment by bollwyvl** [comment]
Have weighed in over on #194, but will reiterate:

- for a "big rock" package like `polars`, translating extras that include a bottom pin into `run-constraints` is likely the right play
  - if it's a _hard_ pin (e.g. `==`) or even a gnarly npm-style `~`, those aren't  going to help folks get to a working environment, and should be isolated... but this takes some finesse
- as on #194, downstreams actually depend on the named extras (which `pip` promptly forgets after being used once, and won't `pip check`) 
  - providing a `noarch: generic` for _each_ `[extra]` until CEP-XXX lands and the entire write/build/lint/install/tick/serve/intaller toolchain supports it, would give humans and blacksmithing robots a way to confidently talk about these things
  - for example:
  ```yaml
  # ...
  requirements:
    run_constrained:
     - pyarrow >=7.0.0
  # ...
  outputs:
    - name: polars
    # ...
    - name: polars-with-pyarrow
      build:
        noarch: generic
      requirements:
        run:
         - pyarrow
  ```

The challenge here is keeping that named set of set of pins up-to-date. As `pip check` will _not_ help, I've used several approaches, but really with `rattler-build`, this is getting a lot easier. On a well-behaved upstreams/recipes this will mean comparing two, well-formed TOML/YAML files instead of layers of `requirements.txt`/`[poetry]` junk and `meta.yml`.

**Comment by pavelzw** [comment]
I personally would prefer waiting for the CEP to come through. Then, ideally `grayskull`/`rattler-build generate-recipe` tooling is also good enough s.t. the dependencies are updated automatically through the autotick bot.
